Server suggest

NBlair52 opened the pull request:
### Basic Suggest logic is in
- If I messed something up when I tried to pull master in let me know. Thought it would make it easier?
- Sometimes there is some weird empty line change/additions (probably windows being dumb)

This still is not complete. Two changes needed
1) (EASY) if a player makes a suggestion, we need to start checking with the next person in line. Right now we just check through the PlayerList (starting with player 1 no matter what, going through all until a refutation is found)
2) (HARDER) need to implement the ability for someone to pick which card to show to refute a suggestion. right now we just check character then room then weapon. If I have two cards that can disprove I should get to pick which one. But this requires some updates to turns tatus etc.

- ADD SuggestRequest - client requesting to server, handles some refuting logic
- ADD RefuteSuggestion - server message to client that another player refuted their suggestion with a card
- ADD SuggestionPassed - server message to all clients that a player was unable to refute a suggestion
- ADD SuggestionWrong - server message to all clients that a player was able to refute a suggestion (no knowledge of how)
- ADD SuggestNotification - server message to all clients that a suggestion was made (and what it is)
- MOVE SolutionHand to Common - this is just a character/room/weapon tupple for me now
- CHANGE PlayerHand - now has isEmpty()
- CHANGE ClueLessClient / ClueLessServer - appropriate changes to suggestion options and handling

NBlair52 committed:
Initial Suggestion Logic and Message Handling

MOVED SolutionHand TO COMMON
Rough First draft.
-For some reason SuggestRequest won't take right now
-Need to add logic to allow players which card to show others if they have the option of two that will disprove the guess
-Need to start checking suggestions against the next player in line, not always starting at the front of the list (easy)

NBlair52 committed:
Merge branch 'master' into Server-Suggest

NBlair52 committed:
Minor fixes to suggestion messaging

-made constructors for SolutionHand public

NBlair52 committed:
Refactoring Suggestion Code

+Added Suggestion UI
~refactored display for Suggestion Messages by putting in toStrings

NBlair52 committed:
Oops - Minor fixes to last

NBlair52 committed:
Updated Server side suggestionhandling

PlayerHand - added isEmpty()
  sometimes throws expception server side
SuggestRequest - added checkRefutation()
  basic logic moved from server to this class to check if a player can 
refute a given suggestion.
ClueLessServer - refactored code out of here and into SuggestRequest

NOTE THIS IS STILL NOT CORRECT. NEED TO UPDATE BY ALLOWING PLAYERS TO 
CHOSE WHICH CARD TO SHOW, AND UPDATE THE ORDER OF CHECKING.

Kit-Kindred commented on Source_Code/Server/ClueLessServer.java:
```diff
@@ -398,20 +407,58 @@ public void processMoveRequest( MoveRequest mr )
+                if ( ! possibleRefutations.isEmpty() )
+                {
+                    sendToAllPlayers( new SuggestionWrong( sr.PlayerName, p.PlayerName ) );
+                    if ( ! possibleRefutations.getCharacters().isEmpty() )
```
The get%CardType%() functions return null rather than an empty collection, so isEmpty() will throw a null pointer exception when there aren't any matching card types. Don't think I have a preference between changing these checks and changing those functions, but one has to change

NBlair52 replied:
changed to check `null`

Kit-Kindred commented on Source_Code/Client/ClueLessClient.java:
```diff
@@ -113,23 +117,23 @@ public static void main(String[] args)
                             {
-                               Thread.sleep( 500 );
-                            } 
+                               Thread.sleep( 1000 );
```
I'm probably going to rewrite a lot of this function with turn handling (and see if there's a way to get rid of this sleep), so maybe you don't need to revert it, but Dave had moved this to 500 ms

NBlair52 replied:
changed back ... sorry

Kit-Kindred commented on Source_Code/Client/ClueLessClient.java:
```diff
@@ -174,29 +178,57 @@ public static void main(String[] args)
                         }
+                case 5 ->
+                        {
+                            System.out.println("Which character would you like to suggest?");
```
I'd add this print for the other two card classes before their prompts too

dav96che replied:
Just weapon probably, since there's no choice of a room card to suggest (it has to be whatever room the player in is)

NBlair52 replied:
The location is going to have to be a TODO, since for right now
1) There is no `Board` being tracked at the server
2) The current `Board` and `BoardRoom `don't have any functionality with the `RoomCard` enum

Kit-Kindred commented on Source_Code/Client/ClueLessClient.java:
```diff
@@ -174,29 +178,57 @@ public static void main(String[] args)
+                            int suggestCharacterIndex = scan.nextInt();
+
+                            index = 1;
+                            for ( RoomName room : RoomName.values() )
```
Let's put a TODO here to make sure we take away the user's ability to select a room once we're keeping track of our current room

NBlair52 replied:
Big TODO left
see above. Need more board functionality before we can make this a reality.

Kit-Kindred commented on Source_Code/Client/ClueLessClient.java:
```diff
@@ -174,29 +178,57 @@ public static void main(String[] args)
+                            }
+                            int suggestWeaponIndex = scan.nextInt();
+
+                            SolutionHand suggestHand = new SolutionHand(CharacterName.values()[ suggestCharacterIndex - 1 ], RoomName.values()[ suggestRoomIndex - 1 ], WeaponType.values()[ suggestWeaponIndex - 1 ]);
```
-1s should really go on the assignment line than here. Not the case right now, but if we were using these indexes for anything else, we'd have to subtract 1 on every use.

NBlair52 replied:
appreciate the tip, changed

Kit-Kindred commented on Source_Code/Common/PlayerHand.java:
```diff
@@ -92,65 +92,82 @@ public void setCard( Card card )
-   
-   
+
+   // retyrbs True if we have no cards
```
Typo

Kit-Kindred commented on Source_Code/Common/SolutionHand.java:
```diff
@@ -1,43 +1,39 @@
+/**
+ *  Solution hand class contains a character, weapon, and room.
+ */
+public class SolutionHand extends PlayerHand
```
I don't think we want to extend PlayerHand for this. This theoretically allows us to add more cards to a SolutionHand, which we don't want. Also seems a bit counterintuitive to have to index into the results of "getPlayerCards" from this class since there should be one and only one card of each type in this object.

NBlair52 replied:
Changed and added more functionality

Kit-Kindred commented on Source_Code/Server/ClueLessServer.java:
```diff
@@ -359,10 +365,13 @@ public void processGameStartRequest(GameStartRequest gsr)
                     CardDeck.shuffleAndAssignCards( PlayerList );
-                    
+
+                    sendToClient( PlayerList.get( CurrentPlayerIndex ).ClientID, new TurnUpdate(true) );
```
Looks like a merge artifact? This is done at the bottom of the function, so I'd remove this turn update unless you deliberately put in an extra

NBlair52 replied:
reluctantly commented out - not sure if this will break anything

Kit-Kindred commented on Source_Code/Client/ClueLessClient.java:
```diff
@@ -319,38 +351,60 @@ else if( statUp instanceof TurnUpdate)
-            
+            //UserPlayer.setHand( ((PlayerHandUpdate) statUp).getHandUpdate() );
+            //System.out.println( UserPlayer.getAllCardsString() + "\n");
+           System.out.println( ((PlayerHandUpdate) statUp).getHandUpdate() + "\n");
```
I think this was also changed in David's most recent commit? If so, should be changed back

Kit-Kindred commented on Source_Code/Common/Messages/ActionRequests/SuggestRequest.java:
```diff
@@ -0,0 +1,73 @@
+        //probably a better way to do this than to loop through
+        for ( CharacterCard playerCard : playerHand.getCharacters())
+        {
+                if (playerCard.getCharacterName() == this.Hand.getCharacters().get(0).getCharacterName())
```
getXName functions currently return strings, so these comparisons should be ...getXName().equals(...getXName()) instead of ==

Although I think those functions should probably be changed to return the enum values instead of strings

NBlair52 replied:
left the legacy "name" functions because Im not sure what removing will break. (I know wrong answer)
added getXEnum and use that now.

Kit-Kindred commented on Source_Code/Server/ClueLessServer.java:
```diff
@@ -359,10 +365,13 @@ public void processGameStartRequest(GameStartRequest gsr)
                     CardDeck.shuffleAndAssignCards( PlayerList );
-                    
+
+                    sendToClient( PlayerList.get( CurrentPlayerIndex ).ClientID, new TurnUpdate(true) );
```
Probably a merge artifact? I don't think we want to do this twice here

Kit-Kindred reviewed (changes requested):
Mostly looks good! Added some comments

NBlair52 committed:
Fixed All of Kits Comments

Added enum methods for each card type
Made SolutionHand not exent PalyerHand
Took choosing room out and added a TODO

Kit-Kindred committed:
Update ClueLessServer.java

Kit-Kindred reviewed (approved):
Had to change ClueLessServer.java:439 back to an isEmpty() check to prevent the first checked player from always generating a refute message
